subsys/bluetooth/host/classic/avdtp.c

```diff
@@ -421,13 +421,102 @@ static void avdtp_get_capabilities_rsp(struct bt_avdtp *session, struct net_buf
 	}
 }
 
+struct bt_avdtp_service_category_handle {
+	uint8_t min_len;
+	uint8_t max_len;
+	bool reconfig_support;
+	uint8_t err_code;
+	uint8_t (*handle)(struct net_buf *buf, uint8_t *len);
+
+};
+
+uint8_t bt_avdtp_check_media_recovery_type(struct net_buf *buf, uint8_t *len)
+{
+	uint8_t recovery_type = net_buf_pull_u8(buf);
+	*len -= sizeof(recovery_type);
+
+	if (recovery_type != BT_AVDTP_RECOVERY_TYPE_FORBIDDEN &&
+		recovery_type != BT_ADVTP_RECOVERY_TYPE_RFC2733) {
+		return BT_AVDTP_BAD_RECOVERY_TYPE;
+	}
+	return BT_AVDTP_SUCCESS;
+}
+static struct bt_avdtp_service_category_handle category_handler[]  = {
+	{0, 0, 0, 0, NULL},
+	{0, 0, 0, BT_AVDTP_BAD_MEDIA_TRANSPORT_FORMAT, NULL},
+	{0, 0, 0, BT_AVDTP_BAD_LENGTH, NULL},
+	{sizeof(struct bt_avdtp_recovery_capabilities),
+	 sizeof(struct bt_avdtp_recovery_capabilities),
+	 0, BT_AVDTP_BAD_RECOVERY_FORMAT, bt_avdtp_check_media_recovery_type},
+	{sizeof(struct bt_avdtp_content_protection_capabilities), UINT8_MAX,
+	 1, BT_AVDTP_BAD_CP_FORMAT, NULL},
+	{sizeof(struct bt_avdtp_header_compression_capabilities),
+	 sizeof(struct bt_avdtp_header_compression_capabilities), 0, BT_AVDTP_BAD_LENGTH, NULL},
+	{sizeof(struct bt_avdtp_multiplexing_capabilities),
+	 sizeof(struct bt_avdtp_multiplexing_capabilities),
+	 0, BT_AVDTP_BAD_MULTIPLEXING_FORMAT, NULL},
+	{sizeof(struct bt_avdtp_media_codec_capabilities), UINT8_MAX, 1, BT_AVDTP_BAD_LENGTH, NULL},
+	{0, 0, 0, BT_AVDTP_BAD_LENGTH, NULL},
+};
+
+uint8_t bt_avdtp_check_service_category(struct net_buf *buf, uint8_t *service_category,
+					bool reconfig)
+{
+	struct bt_avdtp_generic_service_cap *hdr;
+	uint8_t err;
+	struct bt_avdtp_service_category_handle service_handle;
+
+	if (buf->len == 0U) {
+		LOG_DBG("Error: buf not valid");
+		return BT_AVDTP_BAD_LENGTH;
+	}
+
+	while (buf->len > 0U) {
+		if (buf->len < sizeof(*hdr)) {
+			LOG_DBG("Error: buf not valid");
+			return BT_AVDTP_BAD_LENGTH;
+		}
+
+		hdr = net_buf_pull_mem(buf, sizeof(*hdr));
+		*service_category = hdr->service_category;
+
+		if (hdr->service_category != 0 &&
+		    hdr->service_category < ARRAY_SIZE(category_handler)) {
+			service_handle = category_handler[hdr->service_category];
+
+			if (hdr->losc > buf->len || hdr->losc > service_handle.max_len ||
+			    hdr->losc < service_handle.min_len) {
+				return service_handle.err_code;
+			}
+
+			if (!service_handle.reconfig_support && reconfig) {
+				return BT_AVDTP_INVALID_CAPABILITIES;
+			}
+
+			if (service_handle.handle != NULL) {
+				err = service_handle.handle(buf, &(hdr->losc));
```

Comment:
The parameter `len` is not needed. My suggestion is,
```suggestion
				struct net_buf_simple_state state;

				net_buf_simple_save(&buf->b, &state);
				err = service_handle.handle(buf, &(hdr->losc));
				net_buf_simple_restore(&buf->b, &state);
```